### DISCUSSION TITLE
Profiling render times with 'cuda_rgb'

### DISCUSSION THREAD

**Discussion by AakashKT** [mitsuba-renderer/mitsuba3#841] [Q&A] [answered]
Hi,

Thank you for the awesome work with Mitsuba 3!

I was wondering how we can profile render times. 
In my case, I am calling ```mi.render()``` from python three times, for three different integrators.
```
time_ref = time.time()
ref_img = mi.render(scene, integrator='direct' spp=100)
time_ref = time.time() - time_ref

time_1 = time.time()
img_1= mi.render(scene, integrator=integrator_1, spp=100)
time_1= time.time() - time_1

time_2= time.time()
img_2= mi.render(scene, integrator=integrator_2, spp=100)
time_2= time.time() - time_2
```

I know for a fact that ```time_1``` and ```time_2``` should be almost equal, but the latter is larger. In fact, when I evaluate ```integrator_2``` first and then ```integrator_1```, I get opposite effect.

```integrator_1``` and ```integrator_2``` are both witten in python.

Is there a better way to do this profiling? 

Thanks!

**Comment by njroussel** [comment, marked as answer]
Hi @AakashKT 

There's a feature called `KernelHistory` which allows us to get the actual runtimes of the CUDA/Optix kernels. I've given a brief explanation on how to use it in [this comment](https://github.com/mitsuba-renderer/drjit/issues/175#issuecomment-1648025234).

If the kernel runtimes are similiar for the two `mi.render` calls then the only difference is the time it takes to trace through the Python integrator to build the JIT graph.(Note: there should typically be two kernels per `mi.render` call)
If the kernel runtimes differ widely, it most likely has to do with the nature of your integrators. The recommended steps to profile this any deeper would be to remove parts of you integrators to understand the runtime costs that they entail.

**Reply by AakashKT** [reply]
Hi @njroussel ,

Thanks for the quick reply, works like a charm!
